### PR TITLE
[WIP] Add MySQL comment for user creation

### LINKED ISSUE #63237: [FEATURE REQUEST] allow mysql_user.present to specify a comment
**Description**
For documentation purposes, it is nice to have comment on a mysql user to describe why/where/what it is used for. A similar observation can be made for postgresql.

Creating a user in an .sls file could look like
```
frob:
  mysql_user.present:
    - host: 172.0.0.1
    - password: sekret
    - comment: This account is used by the frobnicator service
```

The end query issued to mysql would be
`CREATE USER IF NOT EXISTS 'frob'@'127.0.0.1' identified by 'sekret' COMMENT 'This account is used by the frobnicator service';`

Unfortunately, [this is not implemented](https://docs.saltproject.io/en/latest/ref/states/all/salt.states.mysql_user.html)

### PR DESCRIPTION
**Pull request by itspooya**
### What issues does this PR fix or reference?
Fixes: #63237 

This merge request allows to add comments for mysql user creation

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
- [X] Docs
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
NO

### CODE REVIEW

**Comment by itspooya** [comment]
@twangboy 
Hi Shane,

I just had a couple of questions:

1. Are there any other documentation changes that need to be made in addition to the ones included in this pull request?
2. I noticed that the test on Ubuntu 22.04 failed. Upon reviewing the logs, it looks like this failure was unrelated to the changes made in this pull request (i.e. it was related to Telegram rather than the MySQL/MariaDB modules and states). Is that correct?

```<testcase classname="tests.pytests.unit.beacons.test_telegram_bot_msg" name="test_call_telegram_returning_updates" time="0.003">
<failure message="AssertionError: assert {'photo': [], 'delete_chat_photo': False, 'caption_entities': [], 'new_chat_members': [], 'channel_chat_created': False, 'date': 3155224723, 'group_chat_created': False, 'chat': {'type': 'private', 'username': 'tester', 'id': 1}, 'entities': [], 'supergroup_chat_created': False, 'message_id': 1, 'new_chat_photo': [], 'from': {'username': 'tester', 'is_bot': True, 'id': 1, 'first_name': ''}} == {'photo': [], 'delete_chat_photo': False, 'caption_entities': [], 'new_chat_members': [], 'channel_chat_created': False, 'date': 3155224724, 'group_chat_created': False, 'chat': {'type': 'private', 'username': 'tester', 'id': 1}, 'entities': [], 'supergroup_chat_created': False, 'message_id': 1, 'new_chat_photo': [], 'from': {'username': 'tester', 'is_bot': True, 'id': 1, 'first_name': ''}} Common items: {'caption_entities': [], 'channel_chat_created': False, 'chat': {'id': 1, 'type': 'private', 'username': 'tester'}, 'delete_chat_photo': False, 'entities': [], 'from': {'first_name': '', 'id': 1, 'is_bot': True, 'username': 'tester'}, 'group_chat_created': False, 'message_id': 1, 'new_chat_members': [], 'new_chat_photo': [], 'photo': [], 'supergroup_chat_created': False} Differing items: {'date': 3155224723} != {'date': 3155224724} Full diff: { 'caption_entities': [], 'channel_chat_created': False, 'chat': {'id': 1, 'type': 'private', 'username': 'tester'}, - 'date': 3155224724, ? ^ + 'date': 3155224723, ? ^ 'delete_chat_photo': False, 'entities': [], 'from': {'first_name': '', 'id': 1, 'is_bot': True, 'username': 'tester'}, 'group_chat_created': False, 'message_id': 1, 'new_chat_members': [], 'new_chat_photo': [], 'photo': [], 'supergroup_chat_created': False, }">
def test_call_telegram_returning_updates(): with patch("salt.beacons.telegram_bot_msg.telegram") as telegram_api: token = "abc" username = "tester" config = [{"token": token, "accept_from": [username]}] inst = MagicMock(name="telegram.Bot()") telegram_api.Bot = MagicMock(name="telegram", return_value=inst) user = telegram.User(id=1, first_name="", username=username, is_bot=True) chat = telegram.Chat(1, "private", username=username) date = time.mktime(datetime.datetime(2016, 12, 18, 0, 0).timetuple()) message = telegram.Message(message_id=1, from_user=user, date=date, chat=chat) update = telegram.update.Update(update_id=1, message=message) inst.get_updates.return_value = [update] ret = telegram_bot_msg.validate(config) assert ret == (True, "Valid beacon configuration.") ret = telegram_bot_msg.beacon(config) telegram_api.Bot.assert_called_once_with(token) assert ret > assert ret[0]["msgs"][0] == message.to_dict() E AssertionError: assert {'photo': [], 'delete_chat_photo': False, 'caption_entities': [], 'new_chat_members': [], 'channel_chat_created': False, 'date': 3155224723, 'group_chat_created': False, 'chat': {'type': 'private', 'username': 'tester', 'id': 1}, 'entities': [], 'supergroup_chat_created': False, 'message_id': 1, 'new_chat_photo': [], 'from': {'username': 'tester', 'is_bot': True, 'id': 1, 'first_name': ''}} == {'photo': [], 'delete_chat_photo': False, 'caption_entities': [], 'new_chat_members': [], 'channel_chat_created': False, 'date': 3155224724, 'group_chat_created': False, 'chat': {'type': 'private', 'username': 'tester', 'id': 1}, 'entities': [], 'supergroup_chat_created': False, 'message_id': 1, 'new_chat_photo': [], 'from': {'username': 'tester', 'is_bot': True, 'id': 1, 'first_name': ''}} E Common items: E {'caption_entities': [], E 'channel_chat_created': False, E 'chat': {'id': 1, 'type': 'private', 'username': 'tester'}, E 'delete_chat_photo': False, E 'entities': [], E 'from': {'first_name': '', 'id': 1, 'is_bot': True, 'username': 'tester'}, E 'group_chat_created': False, E 'message_id': 1, E 'new_chat_members': [], E 'new_chat_photo': [], E 'photo': [], E 'supergroup_chat_created': False} E Differing items: E {'date': 3155224723} != {'date': 3155224724} E Full diff: E { E 'caption_entities': [], E 'channel_chat_created': False, E 'chat': {'id': 1, 'type': 'private', 'username': 'tester'}, E - 'date': 3155224724, E ? ^ E + 'date': 3155224723, E ? ^ E 'delete_chat_photo': False, E 'entities': [], E 'from': {'first_name': '', 'id': 1, 'is_bot': True, 'username': 'tester'}, E 'group_chat_created': False, E 'message_id': 1, E 'new_chat_members': [], E 'new_chat_photo': [], E 'photo': [], E 'supergroup_chat_created': False, E } chat = <telegram.chat.Chat object at 0x7fa037fa75e0> config = [{'accept_from': ['tester'], 'token': 'abc'}] date = 1482019200.0 inst = <MagicMock name='telegram.Bot()' id='140326143693616'> message = <telegram.message.Message object at 0x7fa06bce2c40> ret = [{'msgs': [{'caption_entities': [], 'channel_chat_created': False, 'chat': {'id': 1, 'type': 'private', 'username': 'tester'}, 'date': 3155224723, 'delete_chat_photo': False, 'entities': [], 'from': {'first_name': '', 'id': 1, 'is_bot': True, 'username': 'tester'}, 'group_chat_created': False, 'message_id': 1, 'new_chat_members': [], 'new_chat_photo': [], 'photo': [], 'supergroup_chat_created': False}]}] telegram_api = <MagicMock name='telegram' id='140326197055824'> token = 'abc' update = <telegram.update.Update object at 0x7fa03d90e800> user = <telegram.user.User object at 0x7fa03cb65c40> username = 'tester' tests/pytests/unit/beacons/test_telegram_bot_msg.py:137: AssertionError
</failure>
</testcase>```


Let me know if you have any questions or if there is anything else I can do to help. Thanks again for your

**Comment by itspooya** [comment]
This pull request will fix #63237

**Comment by nicholasmhughes** [comment]
re-run pr-windows-2019-x64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-windows-2016-x64-py3-pytest

File: salt/modules/mysql.py [hunk 1]
```diff
@@ -1796,6 +1810,9 @@ def user_create(
     .. versionadded:: 0.16.2
         The ``allow_passwordless`` option was added.
 
+    comments
+        The comments to use for the new user.
```
Comment (nicholasmhughes) [inline salt/modules/mysql.py:1814]: **chore (blocking):** please add a `versionadded` note for this new option

File: salt/states/mysql_user.py [hunk 2]
```diff
@@ -112,6 +113,9 @@ def present(
     unix_socket
         If ``True`` and allow_passwordless is ``True``, the unix_socket auth
         plugin will be used.
+
+    comments
+        comments to add to the user account
```
Comment (nicholasmhughes) [inline salt/states/mysql_user.py:118]: chore (blocking): please add a versionadded note for this new option

File: tests/pytests/unit/states/mysql/test_user.py [hunk 3]
```diff
@@ -60,7 +61,10 @@ def test_present():
         with patch.object(salt.utils.data, "is_true", mock_t):
             comt = "User frank@localhost is already present with passwordless login"
             ret.update({"comment": comt, "result": True})
-            assert mysql_user.present(name, allow_passwordless=True) == ret
+            assert (
```
Comment (nicholasmhughes) [inline tests/pytests/unit/states/mysql/test_user.py:64]: **issue (blocking):** I don't see how the test changes validate anything other than the fact that the state accepts the parameter. Can you add a check for the presence of the proper SQL command in the underlying call args?

**Review by nicholasmhughes — CHANGES_REQUESTED** [review]

File: salt/modules/mysql.py (hunk 1 above)
Comment (itspooya) [reply salt/modules/mysql.py:1814]: What would be the version?

Comment (nicholasmhughes) [reply salt/modules/mysql.py:1814]: realistically, it'll probably be labeled for 3007.0

File: tests/pytests/unit/states/mysql/test_user.py (hunk 3 above)
Comment (itspooya) [reply tests/pytests/unit/states/mysql/test_user.py:64]: I have added a version check to check whether it is using minimum version that supports comment or not

Comment (itspooya) [reply tests/pytests/unit/states/mysql/test_user.py:64]: https://dev.mysql.com/doc/refman/8.0/en/create-user.html#create-user-comments-attributes

**Comment by itspooya** [comment]
Version 8.0.12 was wrong and it is 8.0.21 per document
https://dev.mysql.com/doc/refman/8.0/en/create-user.html#create-user-comments-attributes

**Comment by itspooya** [comment]
A Restart on failed tests would be great

**Comment by itspooya** [comment]
@nicholasmhughes 
Hey Nicholas
I wanted to check with you whether current version is acceptable or more changes needed?

**Comment by nicholasmhughes** [comment]
re-run pr-photon-3-x86_64-py3-pytest

File: tests/pytests/unit/states/mysql/test_user.py (hunk 3 above)
Comment (nicholasmhughes) [reply tests/pytests/unit/states/mysql/test_user.py:64]: I see that you added some version checking in the module code, but I was talking about the pytest tests. You're passing the new `comments` parameter, but there aren't any checks to see if it does what you expect. Perhaps @waynew can take a look during a Test Workshop and give you some ideas on how to test that the underlying call args were modified in a way you'd expect.

**Comment by itspooya** [comment]
How can I re run it?

File: tests/pytests/unit/states/mysql/test_user.py (hunk 3 above)
Comment (itspooya) [reply tests/pytests/unit/states/mysql/test_user.py:64]: Oh now I get it, sorry I thought version checking was enough, I will add appropriate pytest tests, thank you for your comment

**Comment by Ch3LL** [comment]
This still states this PR is a `[wip]` . Is this ready for review now?

**Comment by itspooya** [comment]
Oh Sorry, I forgot about this, I will check it this weekend

**Comment by Ch3LL** [comment]
@itspooya just following up here, is this PR ready for review?

**Pull request closed without merging**